northstar/src/runtime/pipe.rs

```diff
@@ -0,0 +1,256 @@
+                let fd = inner.get_ref().as_raw_fd();
+                // map nix::Error to io::Error
+                match unistd::read(fd, buf.initialized_mut()) {
+                    Ok(n) => Ok(n),
+                    // read(2) on a nonblocking file (O_NONBLOCK) returns EAGAIN or EWOULDBLOCK in
+                    // case that the read would block. That case is handled by `try_io`.
+                    Err(e) => Err(from_nix(e)),
+                }
+            }) {
+                Ok(Ok(n)) => {
+                    buf.advance(n);
+                    return Poll::Ready(Ok(()));
+                }
+                Ok(Err(e)) => {
+                    return Poll::Ready(Err(e));
+                }
+                Err(_would_block) => continue,
+            }
+        }
+    }
+}
+
+/// Pipe's synchronous writing end
+#[derive(Debug)]
+pub(crate) struct AsyncPipeWriter(AsyncFd<PipeWriter>);
+
+impl TryFrom<PipeWriter> for AsyncPipeWriter {
+    type Error = io::Error;
+    fn try_from(writer: PipeWriter) -> Result<Self, Self::Error> {
+        let fd = writer.as_raw_fd();
+        set_nonblocking(fd)?;
+        Ok(AsyncPipeWriter(AsyncFd::new(writer)?))
+    }
+}
+
+impl AsyncWrite for AsyncPipeWriter {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        loop {
+            let mut guard = futures::ready!(self.0.poll_write_ready(cx))?;
+            match guard.try_io(|inner| {
+                let fd = inner.get_ref().as_raw_fd();
+                // map nix::Error to io::Error
+                match unistd::write(fd, buf) {
+                    Ok(n) => Ok(n),
+                    // read(2) on a nonblocking file (O_NONBLOCK) returns EAGAIN or EWOULDBLOCK in
+                    // case that the read would block. That case is handled by `try_io`.
+                    Err(e) => Err(from_nix(e)),
+                }
+            }) {
+                Ok(result) => return Poll::Ready(result),
+                Err(_would_block) => continue,
+            }
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// Sets O_NONBLOCK flag for the input file descriptor
+fn set_nonblocking(fd: RawFd) -> io::Result<()> {
+    let mut flags = fcntl::fcntl(fd, fcntl::FcntlArg::F_GETFL)
+        .map(OFlag::from_bits)
+        .map_err(from_nix)?
+        .unwrap();
+    flags |= OFlag::O_NONBLOCK;
+    fcntl::fcntl(fd, fcntl::FcntlArg::F_SETFL(flags)).map_err(from_nix)?;
+    Ok(())
+}
+
+/// Maps an nix::Error to a io::Error
+fn from_nix(error: nix::Error) -> io::Error {
+    match error {
+        nix::Error::Sys(e) => io::Error::from_raw_os_error(e as i32),
+        e => io::Error::new(io::ErrorKind::Other, format!("{}", e)),
```

Comment:
For `ErrorKind::Other` we can put any error into the `io::Error`. No need to convert it to a `String`.

` e => io::Error::new(io::ErrorKind::Other, e),`